### ISSUE TITLE
sync-notebook needs to look for a Deployment kind that is version independent

### ISSUE THREAD

**Issue by stpabhi** [kubeflow/kubeflow#2185]
k logs po/metacontroller-0 -f
```
I1228 08:46:10.321611       1 request.go:485] Throttling request took 187.307293ms, request: POST:https://10.96.0.1:443/apis/apps/v1beta1/namespaces/resnet50/deployments
I1228 08:46:10.326524       1 controller.go:296] Notebook resnet50/training: child Deployment training created or updated
I1228 08:46:10.339460       1 controller.go:296] Notebook resnet50/training: child Deployment training created or updated
I1228 08:46:10.353557       1 controller.go:296] Notebook resnet50/training: child Deployment training created or updated
I1228 08:46:10.521656       1 request.go:485] Throttling request took 195.589678ms, request: GET:https://10.96.0.1:443/apis/kubeflow.org/v1alpha1/namespaces/resnet50/notebooks/training
I1228 08:46:10.721557       1 request.go:485] Throttling request took 196.969477ms, request: PUT:https://10.96.0.1:443/apis/kubeflow.org/v1alpha1/namespaces/resnet50/notebooks/training
I1228 08:46:10.726439       1 controller.go:388] sync Notebook resnet50/training
I1228 08:46:10.731983       1 manage_children.go:171] Notebook resnet50/training: deleting Deployment resnet50/training
I1228 08:46:10.921371       1 request.go:485] Throttling request took 189.307388ms, request: DELETE:https://10.96.0.1:443/apis/extensions/v1beta1/namespaces/resnet50/deployments/training
I1228 08:46:10.926406       1 manage_children.go:246] Notebook resnet50/training: creating Deployment resnet50/training
I1228 08:46:10.927010       1 controller.go:357] Notebook resnet50/training: child Deployment training deleted
I1228 08:46:10.932837       1 controller.go:357] Notebook resnet50/training: child Deployment training deleted
I1228 08:46:10.937657       1 controller.go:357] Notebook resnet50/training: child Deployment training deleted
I1228 08:46:10.956419       1 controller.go:357] Notebook resnet50/training: child Deployment training deleted
I1228 08:46:11.121330       1 request.go:485] Throttling request took 194.767197ms, request: POST:https://10.96.0.1:443/apis/apps/v1beta1/namespaces/resnet50/deployments
```

**Comment by jlewi** [comment]
@stpabhi Can you elaborate please?  i.e. can you describe what you are concluding based on the logs? 

/cc @kkasravi

**Comment by stpabhi** [comment]
@jlewi I have created a TensorFlow notebook deployment using the notebook.yaml as mentioned here https://github.com/kubeflow/kubeflow/blob/master/kubeflow/jupyter/prototypes/README.md

Deployment, service are created and the notebook pod is running as expected. But the metacontroller is continuously deleting the deployment and recreating it. Above logs shows that notebook deployment is continuously being created and deleted.

**Comment by stpabhi** [comment]
@jlewi Thanks, I’ll do my investigation as well. I’m guessing it might be the status conditions check in sync-notebook.jsonnet and controller couldn’t determine the children status upon creation.

**Comment by kkasravi** [comment]
@stpahbi @jlewi 

Hi - I'll be back on the 6th but will have a bit of time to try and resolve this and push a PR.
I'll explain the issue below:

This is likely due to a version mismatch between the Deployment created 
and the Deployment specified. The logic is as follows:
1. The Deployment will be created in the sync-notebook [code](https://github.com/kubeflow/kubeflow/blob/master/kubeflow/jupyter/sync-notebook.jsonnet#L92)
2. Subsequent polling by the metacontroller will deliver children created in the request object as 
request.children.
3. The sync-notebook [code](https://github.com/kubeflow/kubeflow/blob/master/kubeflow/jupyter/sync-notebook.jsonnet#L118) needs to look for a Deployment kind in request.children that is version independent eg a Deployment created could be apps/v1beta1, apps/v1beta2, apps/v1, extensions/v1beta1 despite specifying apps/v1beta1. If sync-notebook doesn't use what's in request.children - it will be deleted. 
This type of logic is in PR #2154 and is done [here](https://github.com/kubeflow/kubeflow/blob/220d5a98d23ac34c35b56f07401def979878f949/kubeflow/application/sync-application.template#L74) and needs to be done in a similar way. 
4. We likely need a new issue opened so we can push a PR that does 3. above. I think I can get this to you @stpabhi to test before the 6th. @jlewi Let me know what version you would like to target (0.4 RC[X] ....) for this PR. It's not a big change but prevents deletion of resources if the version created doesn't match the version specified (this mostly applies to Deployment since other resources like Service are always "v1".

**Comment by stpabhi** [comment]
@kkasravi Thanks for the detailed explanation. Why does this [here](https://github.com/kubeflow/kubeflow/blob/master/kubeflow/jupyter/notebooks.libsonnet#L130)  have to be ```extensions/v1beta1``` and not ```apps/v1beta1```. Changing it to apps/v1beta1 has fixed the problem for me.

**Comment by kkasravi** [comment]
@stpabhi - yes please do - sorry for the delay

**Comment by jlewi** [comment]
@kkasravi Regarding the release; I don't consider this blocking for 0.4 because I consider it alpha in 0.4. So once we get the appropriate fixes included and cherry-picked into the 0.4 branch we will just include them in the next minor release.

One meta question; is @kkasravi's [comment](https://github.com/kubeflow/kubeflow/issues/2185#issuecomment-450441261) and indication that we should consider switching to implementing the controller in go and not jsonnet? I interpret @kkasravi's comment as an indication that the control logic might need to deal with versioning and this might add sufficient complexity that using an actual programming language would be useful.

**Comment by kkasravi** [comment]
@jlewi switching over to golang is a goal at some point, though right now the logic isn’t overly complex to filter the children passed in as request.children and look for a match with children the controller created. Same code can be leveraged across the existing Metacontrollers.
